### DISCUSSION TITLE
Multiple discriminated unions on same object

### DISCUSSION THREAD

**Discussion by Aaronius** [colinhacks/zod#1006] [Q&A]
Is there a way I can have multiple discriminated unions for the same object? For example, I want to validate an object has a number for `price` if `product` is `shoes`, but a string for `price` if `product` is `shirt` AND that object needs to have a value of `s` or `m` if `color` is `red`, but `m` or `l` if `color` is `blue`.

I'm imagining something like:

```js
const schema = z.discriminatedUnions([
  z.discriminatedUnion("product", [
    z.object({
      product: z.literal("shoes"),
      price: z.number(),
    })
    z.object({
      product: z.literal("shirt"),
      price: z.string()
    })
 ]),
 z.discriminatedUnion("color", [
    z.object({
       color: z.literal("red"),
       size: z.enum(["s", "m"])
    }),
    z.object({
      color: z.literal("blue"),
      size: z.enum(["m", "l"])
    })
]);
```

I've ran into this need multiple times now.

**Comment by knoefel** [comment]
@Aaronius How did you solve this in the meantime? I'm facing the same issue right now.

**Reply by Aaronius** [reply]
I haven't solved it. 😢  I think I'll be coming back to it soon, though. I ended up logging an issue that you may want to thumbs up. https://github.com/colinhacks/zod/issues/1015

**Reply by knoefel** [reply]
That's great, thanks for creating the feature request!

**Reply by Aaronius** [reply]
As a follow-up, I didn't find any good way to solve this. My actual example is more complex than what I posted in my original message. I'm dealing with many properties and, based on the discriminator for the second discriminated union, different properties are supposed to be required. The only way I found of solving this is to run a `superRefine` on the object and manually provide all validation logic for everything the second discriminated union would typically provide. It's gross. If you find anything better, let me know.

**Comment by scotttrinh** [comment]
One possible workaround is to break up the parsing into each piece:

[CodeSandbox](https://codesandbox.io/s/quizzical-https-rid5qg?file=/src/index.ts)
```ts
const productSchema = z.discriminatedUnion("product", [
  z.object({
    product: z.literal("shoes"),
    price: z.number()
  }),
  z.object({
    product: z.literal("shirt"),
    price: z.string()
  })
]);
const colorSchema = z.discriminatedUnion("color", [
  z.object({
    color: z.literal("red"),
    size: z.enum(["s", "m"])
  }),
  z.object({
    color: z.literal("blue"),
    size: z.enum(["m", "l"])
  })
]);

test("valid", (t) => {
  const input = {
    product: "shoes",
    price: 42,
    color: "red",
    size: "s"
  };
  const productResult = productSchema.parse(input);
  const colorResult = colorSchema.parse(input);
  t.deepEqual({ ...productResult, ...colorResult }, input);
  t.end();
});
```

**Reply by knoefel** [reply]
Thanks a lot for your reply! Unfortunately this won't work in my case as i'm using `react-hook-form` and it's `zodResolver` to validate the form. And it expects a single schema for the initialisation.

**Reply by Aaronius** [reply]
Same here.

**Reply by iambryanhaney** [reply]
From your example it looks like a z.union() wrapper is all that's needed. I'm pretty sure this should work (and verified in node REPL that it parses as expected):

```TSX
const schema = z.union([
  z.discriminatedUnion("product", [
    z.object({
      product: z.literal("shoes"),
      price: z.number(),
    }),
    z.object({
      product: z.literal("shirt"),
      price: z.string(),
    }),
  ]),
  z.discriminatedUnion("color", [
    z.object({
      color: z.literal("red"),
      size: z.enum(["s", "m"]),
    }),
    z.object({
      color: z.literal("blue"),
      size: z.enum(["m", "l"]),
    }),
  ]),
]);
```

RHF should handle this fine. Zod should intelligently choose which validations to run based on the discriminator fields' values on submission. After successful validation, the `values` handed to hook-forms `onValid` will still be the union type, but you can switch on the discriminator field values to narrow the exact type.
